sbin/send_email.py

```diff
@@ -12,6 +12,43 @@
 import os
 import datetime
 from sqlalchemy import create_engine, MetaData, Table, bindparam
+import sys
+import psutil
+
+
+# ======================================================================
+#
+# Let's make sure we're the only copy of this script running on the
+# server.  Multiple copies might happen if sending emails is slow or
+# hangs and takes longer than 1 minute and the cron job fires again.
+#
+
+# We could just match the program name, but this is problematic if
+# happens to match the filename submitted by a student.
+# my_program_name = sys.argv[0].split('/')[-1]
+
+# So instead let's match the full path used in the cron script
+my_program_name = sys.argv[0]
+
+my_pid = os.getpid()
+
+# loop over all active processes on the server
+for p in psutil.pids():
+    try:
+        cmdline = psutil.Process(p).cmdline()
+        # if anything on the command line matches the name of the program
+        for i in cmdline:
+            if i.find(my_program_name) != -1:
+                if p != my_pid:
+                    print("ERROR!  Another copy of '" + my_program_name +
+                          "' is already running on the server.  Exiting.")
+                    exit()
```

Comment:
you should use either `sys.exit()` or `raise SystemExit("error message...")` here. The `exit` and `quit` functions shouldn't really be used in a program per the [python docs](https://docs.python.org/3/library/constants.html#constants-added-by-the-site-module).